design: One-pager for Beta Composition Environment, a.k.a. EnvironmentConfigs (#5061)
crossplane/crossplane: design/one-pager-composition-environment-beta.md

# Beta Composition Environment

- Owner: Philippe Scorsolini (@phisco)
- Reviewers: TBD
- Status: Draft

# Background

> The "Composition Environment" concept was introduced in the original

Author: I'll move the old one to defunct once this one gets approved 👍

> The "Composition Environment" concept was introduced in the original

Reviewer: Should we move it to defunct right away or when this one is "implemented"? 

Based on our [conventions](https://github.com/crossplane/crossplane/blob/a53d5f1ff9e0f50e44b421ce5a811251a314437c/design/README.md), I believe the old one should stay `Accepted` with a note that this one supersedes it until its fully implemented, then it should go to `Defunct`? I think this would go into the "semi-defunct" category described at the bottom.

> The "Composition Environment" concept was introduced in the original

Author: yes, I think it makes sense to keep it as semi-defunct, as the parts about `EnvironmentConfigs` will still apply. I'll see what we did in the past and for similar cases and do the same.

    - the lack of visibility of intermediate results for the `in-memory
      environment`
- the `environment`related part of the `Composition` API is perceived as overly
  complex and error-prone, see
  [https://github.com/crossplane/crossplane/issues/4738](https://github.com/crossplane/crossplane/issues/4738).
- having to create an `EnvironmentConfig` to be able to access some information
  from an independent Managed Resource from a `Composition` is considered to be
  cumbersome.

## Promotion to beta

Given all the above, if we imagined splitting the functionality into its
components and promoting them independently, we feel we would be comfortable
promoting to beta the following parts of the `Composition Environment` :

- the `EnvironmentConfig` resource itself
- the `in-memory environment` as an additional source and target for patches
  defined at:
    - `spec.resources[*].patches`
    - `spec.environment.patches`, although these showed some discoverability
      issues and could benefit some more thinking, also keeping into
      consideration these are ignored for function-based Compositions.

While we don’t feel so comfortable promoting the EnvironmentConfig selection

> part of the API at `spec.environment.environmentConfigs` in its current shape.

Author: In this case I mean more like the group of maintainers, I'll make it more explicit.

follows:

![beta-composition-environment-3][beta-composition-environment-3]

### Naming things is hard

Although the initial one-pager and the SIG channel on Slack were named after
the broader concept of `Composition Environment`, since its inception, the
functionality took the name of only one of its parts, `EnvironmentConfig`. This
was even reflected at implementation time by the chosen feature flag,
`--enable-environment-configs`. This caused a lot of confusion as it blurred
the thin line between `EnvironmentConfig` , the resource, and the `in-memory
environment` as a patch source/destination.

Patch types `FromEnvironmentFieldPath` and `ToEnvironmentFieldPath` refer to an
`Environment` which is actually the `in-memory environment`, according to the
definitions above. This `in-memory environment` is decoupled from the selected
`EnvironmentConfig`s and any change to it is not persisted back to any
`EnvironmentConfig` by default. This caused the aforementioned confusion.

As we already saw, beta Composition Functions added support for the `in-memory
environment` by wrapping it into a `Context` object at a well-known key,
`apiextensions.crossplane.io/environment`.

> `spec.environment.environmentConfigs` feels
> [redundant](https://github.com/crossplane/crossplane/issues/4591).

Author: ```suggestion
Referencing the `environment` twice in `spec.environment.environmentConfigs` feels
[redundant](https://github.com/crossplane/crossplane/issues/4591), reinforcing the confusion between the `in-memory environment` and `EnvironmentConfigs`.
```

### Selecting and merging EnvironmentConfigs

The API at `spec.environment.environmentConfigs` grew and now feels
uncomfortably complex. But at the same time, we are
[discussing](https://github.com/crossplane/crossplane/issues/4583) adding even
more complexity to it by adding generic Crossplane resource references to the
`Composition Environment` in some way.

> At the best of our knowledge, mostly power users are selecting and merging

Author: ```suggestion
To the best of our knowledge, it is mostly power users who are selecting and merging
```

[here](https://github.com/crossplane/crossplane/issues/3967) for some related
discussions and possible approaches.

On the other hand, beta composition functions have already improved this aspect
by allowing the very same logic to be run locally using `crossplane beta
render`, possibly running against deployed functions in the near future, and
enabling the injection of arbitrary code at any point in the pipeline through
the usage of a dedicated function, e.g. `function-debug`. Allowing Composition
authors to do "print debugging" as in many other languages. So, we could
postpone any action on this and rely on the improved experience enabled by
Composition Functions, given we are
[proposing](https://github.com/crossplane/crossplane/issues/4746) to deprecate
the native P&T Compositions in the near future.

## Next steps

We currently believe https://github.com/crossplane/crossplane/issues/4739 would
solve all the above issues:

- Independent Managed Resource referencing:
    - functions could request any Crossplane resource, so EnvironmentConfigs when it make sense to, or resources directly when it doesn’t.
- Naming Things:

>     - the concept of `in-memory environment` would just a convention between functions, relying on the [`apiextensions.crossplane.io/environment`](http://apiextensions.crossplane.io/environment) key in the context, as `function-patch-and-transform` already does.

Author: everything else above is conditional, so I'd say:
```suggestion
    - the concept of `in-memory environment` would become just a convention between functions, relying on the [`apiextensions.crossplane.io/environment`](http://apiextensions.crossplane.io/environment) key in the context, as `function-patch-and-transform` already does.
```

However I wouldn't put much focus on the form for this last part, as I want to rewrite it once the above context settles and we agree this is the way to go.

    - `--enable-environment-configs` would really mean only enabling `EnvironmentConfigs` as the rest of the `Composition Environment` would be taken care of by external functions, making the split between the two concepts clearer.
- Selecting and merging EnvironmentConfigs:
    - selection would be done by a dedicated function, e.g. `function-select-extra-resources`.
    - merging into the environment would be done by another dedicated function, e.g. `function-set-environment`.

>     - **N.B.**: how to preserve resolve policy `IfNotPresent` behaviour? could be left to the selection function to implement by adding a place in the composite resource status to store referenced resources.

Author: This was voluntarily left as a question. It's to be discussed whether we care about preserving such behavior first and then how to do that. 
Again don't focus on the form for this section, it's more like a draft proposal to start the discussion as I said in the description of the PR.

>     - **N.B.**: how to preserve resolve policy `IfNotPresent` behaviour? could be left to the selection function to implement by adding a place in the composite resource status to store referenced resources.

Reviewer 2: You mean I was actually meant to read the description of the PR ;-)

>     - **N.B.**: how to preserve resolve policy `IfNotPresent` behaviour? could be left to the selection function to implement by adding a place in the composite resource status to store referenced resources.

Author: I removed the question and moved to the US syntax 😉

  from an independent Managed Resource from a `Composition` is considered to be
  cumbersome.

## Promotion to beta

Given all the above, if we imagined splitting the functionality into its
components and promoting them independently, we, the maintainers, would feel
comfortable promoting to beta the following parts of the `Composition
Environment` :

- the `EnvironmentConfig` resource itself
- the `in-memory environment` as an additional source and target for patches
  defined at:
    - `spec.resources[*].patches`
    - `spec.environment.patches`, although these showed some discoverability
      issues and could benefit some more thinking, also keeping into
      consideration these are ignored for function-based Compositions.

While we wouldn't feel so comfortable promoting the EnvironmentConfig selection
part of the API at `spec.environment.environmentConfigs` in its current shape.

Before promoting the entire feature to beta, it's essential to address the known
issues and make the remaining parts of the Composition Environment more
straightforward.

### Independent Managed Resource referencing

Currently, a Composition author who wanted to address some information from an

> existing Managed Resource would have the following options (simplified a bit

Author: Actually I would rewrite that as:
```
Currently, a Composition author who wanted to address some information external
to the Composite and Composed Resources:
```
the point is that they are "external to the composition", not really whether they are kubernetes or cloud resources. WDYT?

> existing Managed Resource would have the following options (simplified a bit

Author: I rephrased it as above, let me know if there is still something that can be clarified.

the diagram not taking into account what the author knows about the existing
MR):

> ![beta-composition-environment-2][beta-composition-environment-2]

Author: Yes, exactly. It could actually be any known number of resources, the important bit is actually whether you know the resources beforehand or the actual resource to read is controlled by the end user of the composite resource. 

If you know them, today you can create 1 or more EnvironmentConfigs with the needed information and select the right one according to the user's input. 

If you don't know them, you can only compose an ObserveOnly resource according to the user input and therefore create one for each composite resource. 

I'll rephrase it to "Do you need to reference a finite set of Resources you know beforehand?", makes sense?

> ![beta-composition-environment-2][beta-composition-environment-2]

Reviewer: thanks for the clarification. 

I like "Do you need to reference a finite set of Resources you know beforehand?". 

Some other suggestion:
"Do you need to reference a finite set of already known Resources?"

But your suggestion is good as well.

However, all options for existing MRs have some drawbacks:

- A: depending on how many times the Composition is used, could lead to many
  identical ObserveOnly MRs, one for each instantiation
- B: the created EnvironmentConfig is not automatically updated and feels
  unnecessary.
- C: it feels unnecessary, although it helps to have clear and defined
  interfaces between the two compositions; however, it requires some care to
  properly behave in case of multiple Composite resources using the same
  Composition, resulting in multiple EnvironmentConfigs being created.
- D: needs to be actively supported by providers, and there is a will to drop
  it one day to reduce the maintenance burden possibly; usually, they don't

>   cross a single provider's boundaries.

Author: what do you mean by `compiled`?

authors to do "print debugging" as in many other languages. So, we could
postpone any action on this and rely on the improved experience enabled by
Composition Functions, given we are
[proposing](https://github.com/crossplane/crossplane/issues/4746) to deprecate
the native P&T Compositions in the near future.

## Next steps

We currently believe https://github.com/crossplane/crossplane/issues/4739 would
solve all the above issues:

- Independent Managed Resource referencing:
    - functions could request any Crossplane resource, so EnvironmentConfigs when it make sense to, or resources directly when it doesn't.
- Naming Things:
    - the concept of `in-memory environment` would just a convention between functions, relying on the [`apiextensions.crossplane.io/environment`](http://apiextensions.crossplane.io/environment) key in the context, as `function-patch-and-transform` already does.
    - `--enable-environment-configs` would really mean only enabling `EnvironmentConfigs` as the rest of the `Composition Environment` would be taken care of by external functions, making the split between the two concepts clearer.

> - Selecting and merging EnvironmentConfigs:
>     - selection would be done by a dedicated function, e.g. `function-select-extra-resources`.
>     - merging into the environment would be done by another dedicated function, e.g. `function-set-environment`.

Author: sure! `function-patch-and-transform` today only knows about the environment, `function-set-environment` was just an example to make it really explicit it's tied to that, but I would definitely prefer to have it generic from the beginning too 👍
```suggestion
- Selecting and merging EnvironmentConfigs:
    - selection would be done by a dedicated function, e.g. `function-select-extra-resources`.
    - merging into the environment would be done by another dedicated function, e.g. `function-set-environment` or a more generic `function-set-context`.
```

    - **N.B.**: how to preserve resolve policy `IfNotPresent` behaviour? could be left to the selection function to implement by adding a place in the composite resource status to store referenced resources.
- Debugging:
    - visibility already addressed in the dedicated section above
    - selection would be more visible too by allowing passing resources via `--extra-resources` to `crossplane beta render`.

So we would have to:

> - Focus on https://github.com/crossplane/crossplane/issues/4739 for 1.15

Reviewer 3: Just catching up with the latest state there. I need some time to process but my early thoughts are:

I am not yet convinced that the approach there should be the way we support composition environments while it might make sense for requesting extra resources. I feel like we need a more "native" support for environments rather than supporting them as a side effect (of capability of requesting extra resources) with a good amount of boilerplate code and 1-2 function calls each responsible with some part. This is mostly based on the fact that we know composition environments are crucial for any non-trivial composition and the desire to provide a smooth UX to enable this. So, my mind still goes to a solution like [this](https://github.com/crossplane/crossplane/issues/4583#issuecomment-1732343924) instead of splitting this functionality to dedicated functions.

If I try to make an analogy:

We are designing UNIX and discussing how someone should pass some variables into commands and instead of defining/inventing the thing we know as Environment Variables in UNIX, we are talking about suggesting something like the following as a solution for those who need such a functionality:

```shell
env-requester | env-merger | cmd1 | cmd 2
```

instead of:

```shell
export VAR1=value1
source /path/to/somefile
cmd1 | cmd2
```

I know this is not the perfect example, so please don't list the differences between a UNIX environment and K8s/Crossplane etc, the point I am trying to make is more about embracing a widely needed functionality and providing the best possible UX with native support vs "enabling" it somehow.

Author: As per your example, the two things don't exclude each other, in shell scripts, piped commands have access to the environment.

To continue with the UNIX analogy, I see enabling functions to retrieve resources more like providing UNIX sockets and leaving to users to decide what to do with them. 

What I'm suggesting is indeed, given that we have a finite amount of time, to put on hold the development of a native approach and instead focus on enabling functions. Nothing would block us from "bringing back in" as native the "best possible UX" once it emerges from functions and we agree on it.
